xdem/spstats.py

```diff
@@ -0,0 +1,702 @@
+    :param coords: coordinates (if array is 1D)
+    :param range_list: successive ranges with even binning
+    :param nsamp: number of samples to randomly draw from the elevation differences
+    :param nrun: number of samplings
+    :param nproc: number of processing cores
+
+    :return: empirical variogram (variance, lags, counts)
+    """
+
+    # checks
+    dh = dh.squeeze()
+    if coords is None and gsd is None:
+        raise TypeError('Must provide either coordinates or ground sampling distance.')
+    elif gsd is not None and len(dh.shape) == 1:
+        raise TypeError('Array must be 2-dimensional when providing only ground sampling distance')
+    elif coords is not None and len(dh.shape) != 1:
+        raise TypeError('Coordinate array must be provided with 1-dimensional input array')
+    elif coords is not None and (coords.shape[0] != 2 and coords.shape[1] != 2):
+        raise TypeError('One dimension of the coordinates array must be of length equal to 2')
+
+    # defaulting to xx and yy if those are provided
+    if coords is not None:
+        if coords.shape[0] == 2 and coords.shape[1] != 2:
+            coords = np.transpose(coords)
+    else:
+        x, y = np.meshgrid(np.arange(0, dh.shape[0] * gsd, gsd), np.arange(0, dh.shape[1] * gsd, gsd))
+        coords = np.dstack((x.flatten(), y.flatten())).squeeze()
+        dh = dh.flatten()
+
+    # COMMENTING: custom binning is not supported by skgstat yet...
+    # if no range list is specified, define a default one based on the spatial extent of the data and its resolution
+    # if 'bin_func' not in kwargs.keys():
+    #     if range_list is None:
+    #
+    #         # define max range as half the maximum distance between coordinates
+    #         max_range = np.sqrt((np.max(coords[:,0])-np.min(coords[:,0]))**2+(np.max(coords[:,1])-np.min(coords[:,1]))**2)/2
+    #
+    #         # get the ground sampling distance
+    #         if gsd is None:
+    #             est_gsd = np.abs(coords[0,0] - coords[0,1])
+    #         else:
+    #             est_gsd = gsd
+    #
+    #         # define ranges as multiple of the resolution until they get close to the maximum range
+    #         range_list = []
+    #         new_range = gsd
+    #         while new_range < max_range/10:
+    #             range_list.append(new_range)
+    #             new_range *= 10
+    #         range_list.append(max_range)
+    #
+    # else:
+    #     if range_list is not None:
+    #         print('Both range_list and bin_func are defined for binning: defaulting to bin_func')
+
+    # default value we want to use (kmeans is failing)
+    if 'bin_func' not in kwargs.keys():
+        kwargs.update({'bin_func':'even'})
+    if 'n_lags' not in kwargs.keys():
+        kwargs.update({'n_lags':100})
+
+    # estimate variogram
+    if nrun == 1:
+        # subsetting
+        dh_sub, coords_sub = random_subset(dh, coords, nsamp)
+        # getting empirical variogram
+        print(dh_sub.shape)
```

Comment:
The print statement should have some text to explain what is printed.

Reply:
Debug print, removed.